### PR TITLE
fix(core/exec): use nodejs style error messages when throwing

### LINKED ISSUE #16933: go.mod 1.19 update failed with cryptic error message before the go docker image was updated
### How are you running Renovate?

Mend Renovate hosted app on github.com

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the bug

The go.mod update from 1.18 to 1.19 failed in https://github.com/sapcc/keppel/pull/107 because renovate bot was still using a go 1.18 docker image ``docker.io/renovate/go:1.18.5``.
The exact detail can be found in the log ``go: go.mod file indicates go 1.19, but maximum version supported by tidy is 1.18`` but on the PR only a very generic ``Process exited with exit code "1"`` error was logged.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Docker image is already prefetched: docker.io/renovate/go:1.18.5@sha256:e451d2af602fcf158786cb2f1bb1444aefd89768f3c7708a04a768788f6a0a60(branch="renovate/go-1.x")
DEBUG: Executing command(branch="renovate/go-1.x")
{
  "command": "docker run --rm --name=renovate_go --label=renovate_child -v \"/mnt/renovate/gh/sapcc/keppel\":\"/mnt/renovate/gh/sapcc/keppel\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -e GOPATH -e GOPROXY -e GOFLAGS -e CGO_ENABLED -e GIT_CONFIG_KEY_0 -e GIT_CONFIG_VALUE_0 -e GIT_CONFIG_KEY_1 -e GIT_CONFIG_VALUE_1 -e GIT_CONFIG_KEY_2 -e GIT_CONFIG_VALUE_2 -e GIT_CONFIG_COUNT -e GIT_CONFIG_KEY_3 -e GIT_CONFIG_VALUE_3 -e GIT_CONFIG_KEY_4 -e GIT_CONFIG_VALUE_4 -e GIT_CONFIG_KEY_5 -e GIT_CONFIG_VALUE_5 -e BUILDPACK_CACHE_DIR -w \"/mnt/renovate/gh/sapcc/keppel\" docker.io/renovate/go:1.18.5 bash -l -c \"go get -d -t ./... && go mod tidy && go mod vendor && go mod tidy && go mod tidy\""
}
DEBUG: rawExec err(branch="renovate/go-1.x")
{
  "err": {
    "name": "ExecError",
    "cmd": "/bin/sh -c docker run --rm --name=renovate_go --label=renovate_child -v \"/mnt/renovate/gh/sapcc/keppel\":\"/mnt/renovate/gh/sapcc/keppel\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -e GOPATH -e GOPROXY -e GOFLAGS -e CGO_ENABLED -e GIT_CONFIG_KEY_0 -e GIT_CONFIG_VALUE_0 -e GIT_CONFIG_KEY_1 -e GIT_CONFIG_VALUE_1 -e GIT_CONFIG_KEY_2 -e GIT_CONFIG_VALUE_2 -e GIT_CONFIG_COUNT -e GIT_CONFIG_KEY_3 -e GIT_CONFIG_VALUE_3 -e GIT_CONFIG_KEY_4 -e GIT_CONFIG_VALUE_4 -e GIT_CONFIG_KEY_5 -e GIT_CONFIG_VALUE_5 -e BUILDPACK_CACHE_DIR -w \"/mnt/renovate/gh/sapcc/keppel\" docker.io/renovate/go:1.18.5 bash -l -c \"go get -d -t ./... && go mod tidy && go mod vendor && go mod tidy && go mod tidy\"",
    "stderr": "go: downloading github.com/sapcc/go-api-declarations v1.2.0\ngo: downloading github.com/sapcc/go-bits v0.0.0-20220801132244-1778c461b6c3\ngo: downloading github.com/spf13/cobra v1.5.0\ngo: downloading github.com/prometheus/client_golang v1.12.2\ngo: downloading github.com/dlmiddlecote/sqlstats v1.0.2\ngo: downloading github.com/go-redis/redis/v8 v8.11.5\ngo: downloading github.com/gorilla/mux v1.8.0\ngo: downloading github.com/rs/cors v1.8.2\ngo: downloading github.com/docker/distribution v2.8.1+incompatible\ngo: downloading github.com/opencontainers/go-digest v1.0.0\ngo: downloading github.com/gofrs/uuid v4.2.0+incompatible\ngo: downloading gopkg.in/gorp.v2 v2.2.0\ngo: downloading github.com/alicebob/miniredis/v2 v2.22.0\ngo: downloading github.com/go-redis/redis_rate/v9 v9.1.2\ngo: downloading github.com/golang-jwt/jwt/v4 v4.4.2\ngo: downloading golang.org/x/crypto v0.0.0-20220722155217-630584e8d5aa\ngo: downloading github.com/databus23/goslo.policy v0.0.0-20210929125152-81bf2876dbdb\ngo: downloading github.com/gophercloud/gophercloud v0.25.0\ngo: downloading github.com/gophercloud/utils v0.0.0-20220704184730-55bdbbaec4ba\ngo: downloading github.com/majewsky/schwift v1.1.0\ngo: downloading github.com/opencontainers/image-spec v1.0.2\ngo: downloading github.com/inconshreveable/mousetrap v1.0.0\ngo: downloading github.com/spf13/pflag v1.0.5\ngo: downloading github.com/beorn7/perks v1.0.1\ngo: downloading github.com/cespare/xxhash/v2 v2.1.2\ngo: downloading github.com/golang/protobuf v1.5.2\ngo: downloading github.com/prometheus/client_model v0.2.0\ngo: downloading github.com/prometheus/common v0.32.1\ngo: downloading github.com/prometheus/procfs v0.7.3\ngo: downloading golang.org/x/sys v0.0.0-20220412211240-33da011f77ad\ngo: downloading google.golang.org/protobuf v1.28.0\ngo: downloading github.com/dgryski/go-rendezvous v0.0.0-20200823014737-9f7001d12a5f\ngo: downloading github.com/sergi/go-diff v1.2.0\ngo: downloading github.com/lib/pq v1.10.6\ngo: downloading github.com/rabbitmq/amqp091-go v1.3.4\ngo: downloading github.com/alicebob/gopher-json v0.0.0-20200520072559-a9ecdc9d1d3a\ngo: downloading github.com/yuin/gopher-lua v0.0.0-20210529063254-f4c35e4016d9\ngo: downloading gopkg.in/yaml.v2 v2.4.0\ngo: downloading github.com/jpillora/longestcommon v0.0.0-20161227235612-adb9d91ee629\ngo: downloading github.com/hashicorp/golang-lru v0.5.4\ngo: downloading github.com/matttproud/golang_protobuf_extensions v1.0.1\ngo: downloading github.com/hashicorp/go-multierror v1.1.1\ngo: downloading go.uber.org/atomic v1.9.0\ngo: downloading golang.org/x/text v0.3.7\ngo: downloading github.com/mitchellh/go-homedir v1.1.0\ngo: downloading github.com/hashicorp/errwrap v1.0.0\ngo: go.mod file indicates go 1.19, but maximum version supported by tidy is 1.18\n",
    "stdout": "",
    "options": {
      "cwd": "/mnt/renovate/gh/sapcc/keppel",
      "encoding": "utf-8",
      "env": {
        "GOPATH": "/tmp/renovate-cache/others/go",
        "GOPROXY": "https://proxy.golang.org,direct",
        "GOFLAGS": "-modcacherw",
        "CGO_ENABLED": "0",
        "GIT_CONFIG_KEY_0": "url.https://**redacted**@github.com/.insteadOf",
        "GIT_CONFIG_VALUE_0": "ssh://**redacted**@github.com/",
        "GIT_CONFIG_KEY_1": "url.https://**redacted**@github.com/.insteadOf",
        "GIT_CONFIG_VALUE_1": "git@github.com:",
        "GIT_CONFIG_KEY_2": "url.https://**redacted**@github.com/.insteadOf",
        "GIT_CONFIG_VALUE_2": "https://github.com/",
        "GIT_CONFIG_COUNT": "6",
        "GIT_CONFIG_KEY_3": "url.https://**redacted**@api.github.com/.insteadOf",
        "GIT_CONFIG_VALUE_3": "ssh://**redacted**@api.github.com/",
        "GIT_CONFIG_KEY_4": "url.https://**redacted**@api.github.com/.insteadOf",
        "GIT_CONFIG_VALUE_4": "git@api.github.com:",
        "GIT_CONFIG_KEY_5": "url.https://**redacted**@api.github.com/.insteadOf",
        "GIT_CONFIG_VALUE_5": "https://api.github.com/",
        "HOME": "/home/ubuntu",
        "PATH": "/home/ubuntu/.npm-global/bin:/usr/local/node/14.18.1/bin:/home/ubuntu/.local/bin:/usr/local/python/3.9.3/bin:/home/ubuntu/bin:/home/ubuntu/renovateapp/node_modules/.bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
        "LC_ALL": "C.UTF-8",
        "LANG": "C.UTF-8",
        "BUILDPACK_CACHE_DIR": "/tmp/renovate-cache/buildpack"
      },
      "maxBuffer": 10485760,
      "timeout": 900000
    },
    "exitCode": 1,
    "message": "Process exited with exit code \"1\"",
    "stack": "ExecError: Process exited with exit code \"1\"\n    at ChildProcess.<anonymous> (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/exec/common.js:85:24)\n    at ChildProcess.emit (events.js:412:35)\n    at ChildProcess.emit (domain.js:475:12)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:282:12)"
  }
}
```

</details>

### Have you created a minimal reproduction repository?

No reproduction, but I have linked to a public repo where it occurs

### PR DESCRIPTION
**Pull request by Gabriel-Ladzaretti**
## Changes
Use nodejs style error messages when throwing ourselves.
When rethrowing we already reusing the thrown message, so no changes there.

source - 
https://github.com/nodejs/node/blob/7abbbf2ff32b14da47ee4a8e345adb6e424b8689/lib/child_process.js#L409-L415

https://github.com/nodejs/node/blob/7abbbf2ff32b14da47ee4a8e345adb6e424b8689/lib/internal/errors.js#L847-L859

## Context
Closes #16933

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by Gabriel-Ladzaretti** [comment]
https://github.com/ladzaretti/exec-error-message/pull/2
https://github.com/ladzaretti/exec-error-message/pull/1

might as well close #16916 and make https://github.com/renovatebot/renovate/pull/16917 redundant

**Comment by rarkins** [comment]
Can you compare this to what we had previously when using exec?

**Comment by Gabriel-Ladzaretti** [comment]
looks the same for go
note that go mod version is not supported in the version i ran 32.94.0
https://github.com/ladzaretti/exec-error-message-pre-spawn/pull/1

as for composer, the structure is the same, but i dont have composer installed locally therefore the stderr is different 
https://github.com/ladzaretti/composer-testing-pre-spawn/pull/2

**Comment by Gabriel-Ladzaretti** [comment]
as for 
- https://github.com/renovatebot/renovate/issues/16916

there is still not standardization, and some managers pring error.message and append stderr with will result in duplicates, some even append stdout.

but with this PR we are basically in the same situation as we were before the spawn refactor (with artifacts error logs in PRs)

File: lib/util/exec/common.spec.ts [hunk 1]
```diff
@@ -200,7 +200,7 @@ describe('util/exec/common', () => {
       ).rejects.toMatchObject({
         cmd,
         signal: exitSignal,
-        message: 'Process signaled with "SIGTERM"',
+        message: `Command failed: ${cmd}\n`,
```
Comment (rarkins) [inline lib/util/exec/common.spec.ts:203]: Can we indicate it was preempted by SIGTERM?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/util/exec/common.spec.ts (hunk 1 above)
Comment (Gabriel-Ladzaretti) [reply lib/util/exec/common.spec.ts:203]: yeah, the execError has signals and codes if available. 
only the `error.message` is changed
see exitCode below
```
DEBUG: Failed to generate composer.lock (repository=ladzaretti/composer-testing, branch=renovate/monolog-monolog-3.x)
       "err": {
         "name": "ExecError",
         "cmd": "C:\\Windows\\system32\\cmd.exe /d /s /c \"composer update monolog/monolog --with-dependencies --ignore-platform-req='ext-*' --ignore-platform-req='lib-*' --no-ansi -
-no-interaction --no-scripts --no-autoloader --no-plugins\"",
         "stderr": "'composer' is not recognized as an internal or external command,\r\noperable program or batch file.\r\n",
         "stdout": "",
         "options": {
           "cwd": "C:/Users/GABRIE~1/AppData/Local/Temp/renovate/repos/github/ladzaretti/composer-testing",
           "encoding": "utf-8",
           "env": {
             "COMPOSER_CACHE_DIR": "C:/Users/GABRIE~1/AppData/Local/Temp/renovate/cache/others/composer",
             "COMPOSER_AUTH": "{\"github-oauth\":{\"github.com\":\"**redacted**\"}}",
             "PATH": "C:\\Users\\GABRIE~1\\AppData\\Local\\Temp\\yarn--1659602435390-0.17226112613140776;C:\\projects\\ladzaretti-renovate\\node_modules\\.bin;C:\\Users\\GabrielLadza
retti\\AppData\\Local\\Yarn\\Data\\link\\node_modules\\.bin;C:\\Users\\GabrielLadzaretti\\AppData\\Local\\Yarn\\bin;C:\\Program Files\\libexec\\lib\\node_modules\\npm\\bin\\node-gyp-
bin;C:\\Program Files\\lib\\node_modules\\npm\\bin\\node-gyp-bin;C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\node-gyp-bin;C:\\Users\\GabrielLadzaretti\\Downloads\\jdk-11\\bin;
C:\\Program Files (x86)\\Common Files\\Oracle\\Java\\javapath;C:\\Windows\\system32;C:\\Windows;C:\\Windows\\System32\\Wbem;C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\;C:\\Windo
ws\\System32\\OpenSSH\\;C:\\Program Files\\dotnet\\;C:\\apache-maven-3.8.4-bin\\apache-maven-3.8.4\\bin;C:\\Program Files\\vscode;C:\\Gradle\\gradle-7.3.3\\bin;C:\\Program Files\\Git
\\cmd;C:\\Program Files\\nodejs\\;C:\\Program Files\\Docker\\Docker\\resources\\bin;C:\\ProgramData\\DockerDesktop\\version-bin;C:\\Program Files\\Go\\bin;C:\\Users\\GabrielLadzarett
i\\AppData\\Local\\Microsoft\\WindowsApps;C:\\Program Files\\JetBrains\\IntelliJ IDEA 2021.3.1\\bin;;C:\\Program Files\\JetBrains\\GoLand 2021.3.2\\bin;;C:\\Program Files\\JetBrains\
\WebStorm 2021.3.1\\bin;;C:\\Users\\GabrielLadzaretti\\AppData\\Roaming\\npm;C:\\Users\\GabrielLadzaretti\\go\\bin"
           },
           "maxBuffer": 10485760,
           "timeout": 900000
         },
         "exitCode": 1,
         "message": "Command failed: composer update monolog/monolog --with-dependencies --ignore-platform-req='ext-*' --ignore-platform-req='lib-*' --no-ansi --no-interaction --no-s
cripts --no-autoloader --no-plugins\n'composer' is not recognized as an internal or external command,\r\noperable program or batch file.\r\n",
         "stack": "ExecError: Command failed: composer update monolog/monolog --with-dependencies --ignore-platform-req='ext-*' --ignore-platform-req='lib-*' --no-ansi --no-interacti
on --no-scripts --no-autoloader --no-plugins\n'composer' is not recognized as an internal or external command,\r\noperable program or batch file.\r\n\n    at ChildProcess.<anonymous>
 (C:\\projects\\ladzaretti-renovate\\lib\\util\\exec\\common.ts:94:16)\n    at ChildProcess.emit (node:events:402:35)\n    at ChildProcess.emit (node:domain:475:12)\n    at Process.C
hildProcess._handle.onexit (node:internal/child_process:290:12)"
       }

```

Comment (Gabriel-Ladzaretti) [reply lib/util/exec/common.spec.ts:203]: image

Comment (Gabriel-Ladzaretti) [reply lib/util/exec/common.spec.ts:203]: but i guess its available in the logs only

Comment (Gabriel-Ladzaretti) [reply lib/util/exec/common.spec.ts:203]: misunderstood first, i guess for signals we can have a slightly different message which includes the signal?

but regardless both signal and exit codes are available in the logs.
should we add the signal to the error message?

Comment (rarkins) [reply lib/util/exec/common.spec.ts:203]: I think it's valuable to indicate that it didn't fail due to the package manager but instead due to being interrupted. Even if it's `Command failed ${cmd}\nInterrupted by SIGTERM`

Comment (rarkins) [reply lib/util/exec/common.spec.ts:203]: We could maybe even abort this branch in future btw so that it can be retried on the next run

Comment (rarkins) [reply lib/util/exec/common.spec.ts:203]: Or retry it immediately if that's elegant to do

Comment (Gabriel-Ladzaretti) [reply lib/util/exec/common.spec.ts]: message changed to `Command failed ${cmd}\nInterrupted by ${signal}`, i omitted the stderr from the message, as it not quite helpful in this context, so does stdout. 
but still, both available in the logs though

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Comment by Gabriel-Ladzaretti** [comment]
so if the branch is out of date with main, it wont get merged, makes sense 😅

**Pull request merged**